parquet-column/src/main/java/org/apache/parquet/column/values/factory/ValuesWriterFactory.java

```diff
@@ -0,0 +1,40 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.parquet.column.values.factory;
+
+import org.apache.parquet.column.ColumnDescriptor;
+import org.apache.parquet.column.values.ValuesWriter;
+
+/**
+ * Can be overridden to allow users to specify how they want their ValuesWriters to be created.
+ * ValuesWriterFactories are created using reflection in {@link org.apache.parquet.column.ParquetProperties}.
+ * ValuesWriterFactories can in turn read additional config to create appropriate ValuesWriters.
+ */
+public interface ValuesWriterFactory {
+
+  /**
+   * Used to initialize the factory. This method is called before newValuesWriter()
+   */
+  void initialize(ValuesWriterFactoryParams params);
```

Comment:
Yeah so `initialize(...)` is called only once. `newValuesWriter(...)` is called once per column per block. So basically we instantiate a new ValuesWriter for each column at the start of the block (and repeat on new blocks). 

As we have kept this approach open ended, it is up to folks how they do things. We could have the factory hold on to state if needed (or call out to a metadata service that holds metadata about the ideal encoding for a given column). We could also have scenarios where the ValuesWriter itself is intelligent. For example if we want to have a valuesWriter evaluate a few encoding strategies and fallback to the best one (similar to the existing FallbackWriter). 

I'll document the lifecycle of the ValuesWriter object so that folks implementing it have a good sense of what constraints they have to work with.